File: ml-platform/backend/app/models/model_registry.py
"""Immutable model registry versions and inference deployments."""

import uuid

from sqlalchemy import (
    JSON,
    CheckConstraint,
    Column,
    DateTime,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
    UniqueConstraint,
    func,
    text,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from app.database import Base


MODEL_SOURCE_KINDS = ("platform_joblib", "onnx_artifact")
APPROVAL_STATES = ("pending", "approved", "rejected", "archived")
DESIRED_STATES = ("stopped", "running")
OBSERVED_STATES = ("stopped", "starting", "running", "stopping", "failed")
REVISION_STRATEGIES = ("immediate", "canary", "rolling")
REVISION_STATES = ("draft", "candidate", "stable", "superseded", "failed")
ROLLOUT_STATES = (
    "pending",
    "preloading",
    "progressing",
    "paused",
    "completed",
    "failed",
    "rolled_back",
)


class RegisteredModel(Base):
    __tablename__ = "registered_models"
    __table_args__ = (
        UniqueConstraint(
            "project_id",
            "name",
            name="uq_registered_models_project_name",
        ),
        Index(
            "ix_registered_models_project_created",
            "project_id",
            "created_at",
        ),
    )

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    project_id = Column(
        UUID(as_uuid=True),
        ForeignKey("projects.id", ondelete="CASCADE"),
        nullable=False,
    )
    name = Column(String(128), nullable=False)
    description = Column(Text, nullable=False, default="")
    created_by_id = Column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )
    created_at = Column(DateTime, nullable=False, server_default=func.now())
    updated_at = Column(
        DateTime,
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )

    versions = relationship(
        "ModelVersion",
        back_populates="registered_model",
        cascade="all, delete-orphan",
        passive_deletes=True,
    )


class ModelVersion(Base):
    __tablename__ = "model_versions"
    __table_args__ = (
        UniqueConstraint(
            "registered_model_id",
            "version_number",
            name="uq_model_versions_model_number",
        ),
        CheckConstraint(
            "source_kind IN ('platform_joblib', 'onnx_artifact')",
            name="ck_model_versions_source_kind",
        ),
        CheckConstraint(
            "approval_status IN ('pending', 'approved', 'rejected', 'archived')",
            name="ck_model_versions_approval_status",
        ),
        CheckConstraint(
            "(source_kind = 'platform_joblib' AND source_model_library_id IS NOT NULL) "
            "OR (source_kind = 'onnx_artifact' AND source_model_library_id IS NULL)",
            name="ck_model_versions_source_reference",
        ),
        CheckConstraint(
            "version_number > 0",
            name="ck_model_versions_positive_number",
        ),
        Index(
            "ix_model_versions_model_created",
            "registered_model_id",
            "created_at",
        ),
        Index(
            "ix_model_versions_approval_created",
            "approval_status",
            "created_at",
        ),
    )

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    registered_model_id = Column(
        UUID(as_uuid=True),
        ForeignKey("registered_models.id", ondelete="CASCADE"),
        nullable=False,
    )
    version_number = Column(Integer, nullable=False)
    source_kind = Column(String(32), nullable=False)
    source_model_library_id = Column(
        UUID(as_uuid=True),
        ForeignKey("model_library.id", deferrable=True, initially="DEFERRED"),
        nullable=True,
    )
    source_artifact_id = Column(
        UUID(as_uuid=True),
        ForeignKey("artifacts.id", deferrable=True, initially="DEFERRED"),
        nullable=False,
    )
    onnx_artifact_id = Column(
        UUID(as_uuid=True),
        ForeignKey("artifacts.id", deferrable=True, initially="DEFERRED"),
        nullable=False,
    )
    framework = Column(String(64), nullable=False, default="")
    algorithm = Column(String(128), nullable=False, default="")
    feature_schema = Column(JSON, nullable=False, default=list)
    output_schema = Column(JSON, nullable=False, default=dict)
    metrics = Column(JSON, nullable=False, default=dict)
    conversion_metadata = Column(JSON, nullable=False, default=dict)
    approval_status = Column(String(16), nullable=False, default="pending")
    approval_comment = Column(Text, nullable=False, default="")
    approved_by_id = Column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )
    approved_at = Column(DateTime, nullable=True)
    created_by_id = Column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )
    created_at = Column(DateTime, nullable=False, server_default=func.now())

    registered_model = relationship("RegisteredModel", back_populates="versions")


class InferenceDeployment(Base):
    __tablename__ = "inference_deployments"
    __table_args__ = (
        UniqueConstraint(
            "project_id",
            "name",
            name="uq_inference_deployments_project_name",
        ),
        CheckConstraint(
            "desired_state IN ('stopped', 'running')",
            name="ck_inference_deployments_desired_state",
        ),
        CheckConstraint(
            "observed_state IN ('stopped', 'starting', 'running', 'stopping', 'failed')",
            name="ck_inference_deployments_observed_state",
        ),
        Index(
            "ix_inference_deployments_project_state",
            "project_id",
            "observed_state",
        ),
        Index(
            "ix_inference_deployments_desired_checked",
            "desired_state",
            "last_checked_at",
        ),
    )

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    project_id = Column(
        UUID(as_uuid=True),
        ForeignKey("projects.id", ondelete="CASCADE"),
        nullable=False,
    )
    name = Column(String(128), nullable=False)
    model_version_id = Column(
        UUID(as_uuid=True),
        ForeignKey("model_versions.id", deferrable=True, initially="DEFERRED"),
        nullable=False,
    )
    desired_state = Column(String(16), nullable=False, default="stopped")
    observed_state = Column(String(16), nullable=False, default="stopped")
    last_error_code = Column(String(64), nullable=True)
    last_checked_at = Column(DateTime, nullable=True)
    started_at = Column(DateTime, nullable=True)
    stopped_at = Column(DateTime, nullable=True)
    created_by_id = Column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )
    created_at = Column(DateTime, nullable=False, server_default=func.now())
    updated_at = Column(
        DateTime,
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )

    model_version = relationship("ModelVersion")


class DeploymentRevision(Base):
    __tablename__ = "deployment_revisions"
    __table_args__ = (
        UniqueConstraint(
            "deployment_id",
            "revision_number",
            name="uq_deployment_revisions_number",
        ),
        CheckConstraint(
            "revision_number > 0",
            name="ck_deployment_revisions_positive_number",
        ),
        CheckConstraint(
            "strategy IN ('immediate', 'canary', 'rolling')",
            name="ck_deployment_revisions_strategy",
        ),
        CheckConstraint(
            "status IN ('draft', 'candidate', 'stable', 'superseded', 'failed')",
            name="ck_deployment_revisions_status",
        ),
        Index(
            "ix_deployment_revisions_deployment_status",
            "deployment_id",
            "status",
        ),
    )

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    deployment_id = Column(
        UUID(as_uuid=True),
        ForeignKey("inference_deployments.id", ondelete="CASCADE"),
        nullable=False,
    )
    revision_number = Column(Integer, nullable=False)
    strategy = Column(String(16), nullable=False)
    status = Column(String(16), nullable=False, default="draft")
    created_by_id = Column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )
    created_at = Column(DateTime, nullable=False, server_default=func.now())
    activated_at = Column(DateTime, nullable=True)

    deployment = relationship("InferenceDeployment")
    targets = relationship(
        "DeploymentTarget",
        back_populates="revision",
        cascade="all, delete-orphan",
        passive_deletes=True,
    )


class DeploymentTarget(Base):
    __tablename__ = "deployment_targets"
    __table_args__ = (
        UniqueConstraint(
            "revision_id",
            "model_version_id",
            name="uq_deployment_targets_revision_model",
        ),
        CheckConstraint(
            "weight_bps >= 0 AND weight_bps <= 10000",
            name="ck_deployment_targets_weight",
        ),
        CheckConstraint(
            "role IN ('stable', 'candidate')",
            name="ck_deployment_targets_role",
        ),
        Index("ix_deployment_targets_model_version", "model_version_id"),
    )

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    revision_id = Column(
        UUID(as_uuid=True),
        ForeignKey("deployment_revisions.id", ondelete="CASCADE"),
        nullable=False,
    )
    model_version_id = Column(
        UUID(as_uuid=True),
        ForeignKey("model_versions.id", deferrable=True, initially="DEFERRED"),
        nullable=False,
    )
    weight_bps = Column(Integer, nullable=False)
    role = Column(String(16), nullable=False)

    revision = relationship("DeploymentRevision", back_populates="targets")
    model_version = relationship("ModelVersion")


class DeploymentRollout(Base):
    __tablename__ = "deployment_rollouts"
    __table_args__ = (
        CheckConstraint(
            "state IN ('pending', 'preloading', 'progressing', 'paused', "
            "'completed', 'failed', 'rolled_back')",
            name="ck_deployment_rollouts_state",
        ),
        CheckConstraint(
            "current_step >= 0",
            name="ck_deployment_rollouts_current_step",
        ),
        CheckConstraint(
            "lock_version > 0",
            name="ck_deployment_rollouts_lock_version",
        ),
        Index(
            "uq_deployment_rollouts_active",
            "deployment_id",
            unique=True,
            postgresql_where=text(
                "state IN ('pending', 'preloading', 'progressing', 'paused')"
            ),
            sqlite_where=text(
                "state IN ('pending', 'preloading', 'progressing', 'paused')"
            ),
        ),
        Index(
            "ix_deployment_rollouts_deployment_created",
            "deployment_id",
            "created_at",
        ),
    )

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    deployment_id = Column(
        UUID(as_uuid=True),
        ForeignKey("inference_deployments.id", ondelete="CASCADE"),
        nullable=False,
    )
    from_revision_id = Column(
        UUID(as_uuid=True),
        ForeignKey(
            "deployment_revisions.id",
            deferrable=True,
            initially="DEFERRED",
        ),
        nullable=True,
    )
    to_revision_id = Column(
        UUID(as_uuid=True),
        ForeignKey(
            "deployment_revisions.id",
            deferrable=True,
            initially="DEFERRED",
        ),
        nullable=False,
    )
    state = Column(String(16), nullable=False, default="pending")
    current_step = Column(Integer, nullable=False, default=0)
    lock_version = Column(Integer, nullable=False, default=1)
    step_schedule = Column(JSON, nullable=False, default=list)
    thresholds = Column(JSON, nullable=False, default=dict)
    last_error_code = Column(String(64), nullable=True)
    created_at = Column(DateTime, nullable=False, server_default=func.now())
    started_at = Column(DateTime, nullable=True)
    updated_at = Column(
        DateTime,
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )
    completed_at = Column(DateTime, nullable=True)

    deployment = relationship("InferenceDeployment")
    from_revision = relationship(
        "DeploymentRevision",
        foreign_keys=[from_revision_id],
    )
    to_revision = relationship(
        "DeploymentRevision",
        foreign_keys=[to_revision_id],
    )


class InferenceApiKey(Base):
    __tablename__ = "inference_api_keys"
    __table_args__ = (
        CheckConstraint(
            "length(prefix) = 12",
            name="ck_inference_api_keys_prefix_length",
        ),
        Index(
            "ix_inference_api_keys_deployment_prefix",
            "deployment_id",
            "prefix",
            unique=True,
        ),
    )

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    deployment_id = Column(
        UUID(as_uuid=True),
        ForeignKey("inference_deployments.id", ondelete="CASCADE"),
        nullable=False,
    )
    prefix = Column(String(12), nullable=False)
    secret_hash = Column(String(512), nullable=False)
    scopes = Column(JSON, nullable=False, default=list)
    expires_at = Column(DateTime, nullable=True)
    revoked_at = Column(DateTime, nullable=True)
    last_used_at = Column(DateTime, nullable=True)
    created_by_id = Column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )
    created_at = Column(DateTime, nullable=False, server_default=func.now())

    deployment = relationship("InferenceDeployment")


class InferenceRequestLog(Base):
    __tablename__ = "inference_request_logs"
    __table_args__ = (
        UniqueConstraint(
            "request_id",
            name="uq_inference_request_logs_request_id",
        ),
        CheckConstraint(
            "status IN ('success', 'error', 'limited')",
            name="ck_inference_request_logs_status",
        ),
        Index(
            "ix_inference_request_logs_deployment_occurred",
            "deployment_id",
            "occurred_at",
        ),
        Index("ix_inference_request_logs_expires", "expires_at"),
    )

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    request_id = Column(String(128), nullable=False)
    deployment_id = Column(
        UUID(as_uuid=True),
        ForeignKey("inference_deployments.id", ondelete="CASCADE"),
        nullable=False,
    )
    revision_id = Column(
        UUID(as_uuid=True),
        ForeignKey("deployment_revisions.id", ondelete="SET NULL"),
        nullable=True,
    )
    model_version_id = Column(
        UUID(as_uuid=True),
        ForeignKey("model_versions.id", ondelete="SET NULL"),
        nullable=True,
    )
    api_key_id = Column(
        UUID(as_uuid=True),
        ForeignKey("inference_api_keys.id", ondelete="SET NULL"),
        nullable=True,
    )
    batch_size = Column(Integer, nullable=False)
    duration_ms = Column(Integer, nullable=False)
    status = Column(String(16), nullable=False, default="success")
    error_code = Column(String(64), nullable=True)
    occurred_at = Column(DateTime, nullable=False)
    expires_at = Column(DateTime, nullable=False)

    deployment = relationship("InferenceDeployment")
    revision = relationship("DeploymentRevision")
    model_version = relationship("ModelVersion")
    api_key = relationship("InferenceApiKey")


class InferenceMetricBucket(Base):
    __tablename__ = "inference_metric_buckets"
    __table_args__ = (
        Index(
            "uq_inference_metric_buckets_deployment_minute",
            "deployment_id",
            "bucket_start",
            unique=True,
        ),
    )

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    deployment_id = Column(
        UUID(as_uuid=True),
        ForeignKey("inference_deployments.id", ondelete="CASCADE"),
        nullable=False,
    )
    bucket_start = Column(DateTime, nullable=False)
    request_count = Column(Integer, nullable=False, default=0)
    success_count = Column(Integer, nullable=False, default=0)
    error_count = Column(Integer, nullable=False, default=0)
    limited_count = Column(Integer, nullable=False, default=0)
    load_failure_count = Column(Integer, nullable=False, default=0)
    batch_size_sum = Column(Integer, nullable=False, default=0)
    latency_sum_ms = Column(Integer, nullable=False, default=0)
    latency_max_ms = Column(Integer, nullable=False, default=0)
    latency_buckets = Column(JSON, nullable=False, default=dict)
    traffic_weights = Column(JSON, nullable=False, default=dict)

    deployment = relationship("InferenceDeployment")


class ModelCard(Base):
    __tablename__ = "model_cards"
    __table_args__ = (
        UniqueConstraint(
            "model_version_id",
            name="uq_model_cards_model_version",
        ),
        Index("ix_model_cards_release_status", "release_status"),
    )

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    model_version_id = Column(
        UUID(as_uuid=True),
        ForeignKey("model_versions.id", ondelete="CASCADE"),
        nullable=False,
    )
    training_data_lineage = Column(JSON, nullable=False, default=dict)
    source_artifact_ids = Column(JSON, nullable=False, default=list)
    input_schema = Column(JSON, nullable=False, default=list)
    output_schema = Column(JSON, nullable=False, default=dict)
    metrics = Column(JSON, nullable=False, default=dict)
    approval_history = Column(JSON, nullable=False, default=list)
    approval_status = Column(String(16), nullable=False, default="pending")
    release_status = Column(String(16), nullable=False, default="unreleased")
    risk_notes = Column(Text, nullable=False, default="")
    intended_use = Column(String(4000), nullable=False, default="")
    limitations = Column(String(4000), nullable=False, default="")
    operational_guidance = Column(Text, nullable=False, default="")
    guidance_revision = Column(Integer, nullable=False, default=1)
    created_at = Column(DateTime, nullable=False, server_default=func.now())
    updated_at = Column(
        DateTime,
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )

    model_version = relationship("ModelVersion")

File: ml-platform/frontend/src/api/modelRegistry.ts
import apiClient from "./client";

export type ProjectRole = "owner" | "editor" | "operator" | "viewer";
export type ModelSourceKind = "platform_joblib" | "onnx_artifact";
export type ApprovalStatus = "pending" | "approved" | "rejected" | "archived";
export type DesiredState = "stopped" | "running";
export type ObservedState = "stopped" | "starting" | "running" | "stopping" | "failed";

export interface ProjectOption {
  id: string;
  name: string;
  project_role: ProjectRole;
}

export interface FeatureField {
  name: string;
  dtype: string;
}

export interface OutputSchema {
  name: string;
  dtype: string;
  task: "classification" | "regression";
}

export interface RegisteredModel {
  id: string;
  project_id: string;
  name: string;
  description: string;
  latest_version: number | null;
  latest_approval_status: ApprovalStatus | null;
  created_at: string | null;
}

export interface ModelVersion {
  id: string;
  registered_model_id: string;
  version_number: number;
  source_kind: ModelSourceKind;
  framework: string;
  algorithm: string;
  feature_schema: FeatureField[];
  output_schema: OutputSchema;
  metrics: Record<string, number | string | boolean | null>;
  conversion_metadata: Record<string, unknown>;
  approval_status: ApprovalStatus;
  approval_comment: string;
  created_at: string | null;
}

export interface InferenceDeployment {
  id: string;
  project_id: string;
  name: string;
  model_version_id: string;
  desired_state: DesiredState;
  observed_state: ObservedState;
  last_error_code: string | null;
  last_checked_at: string | null;
  created_at: string | null;
}

export type RecordValue = string | number | boolean;
export type InferenceRecord = Record<string, RecordValue>;

export interface PredictionResult {
  deployment_id: string;
  model_version_id: string;
  version_number: number;
  predictions: unknown[];
  probabilities?: unknown[];
  duration_ms: number;
}

export type RolloutStrategy = "immediate" | "canary" | "rolling";
export type RolloutState = "pending" | "preloading" | "progressing" | "paused" | "completed" | "failed" | "rolled_back";
export type RevisionStatus = "draft" | "candidate" | "stable" | "superseded" | "failed";

export interface DeploymentTarget {
  model_version_id: string;
  weight_bps: number;
}

export interface DeploymentRevision {
  id: string;
  deployment_id: string;
  revision_number: number;
  strategy: RolloutStrategy;
  status: RevisionStatus;
  targets: DeploymentTarget[];
  created_at: string | null;
  activated_at: string | null;
}

export interface DeploymentRollout {
  id: string;
  deployment_id: string;
  from_revision_id: string | null;
  to_revision_id: string;
  state: RolloutState;
  current_step: number;
  lock_version: number;
  step_schedule: number[];
  thresholds: Record<string, number>;
  last_error_code: string | null;
  created_at: string | null;
  started_at: string | null;
  completed_at: string | null;
  targets: DeploymentTarget[];
}

export interface InferenceApiKey {
  id: string;
  prefix: string;
  scopes: string[];
  expires_at: string | null;
  last_used_at: string | null;
  revoked_at: string | null;
  created_at: string | null;
}

export interface CreatedInferenceApiKey extends InferenceApiKey {
  plaintext: string;
}

export interface InferenceMetricBucket {
  bucket_start: string;
  request_count: number;
  success_count: number;
  error_count: number;
  limited_count: number;
  load_failure_count: number;
  latency_buckets: Record<string, number>;
  traffic_weights: Record<string, number>;
}

export interface InferenceMetricSummary {
  request_count: number;
  success_count?: number;
  error_count: number;
  limited_count?: number;
  load_failure_count?: number;
  average_batch_size?: number;
  average_latency_ms?: number;
  max_latency_ms?: number;
  p50_latency_ms?: number | null;
  p95_latency_ms?: number | null;
  p99_latency_ms?: number | null;
  latency_buckets?: Record<string, number>;
  traffic_weights?: Record<string, number>;
}

export interface InferenceMetricPage {
  items: InferenceMetricBucket[];
  summary: InferenceMetricSummary;
  page: number;
  page_size: number;
}

export interface InferenceRequestLog {
  id: string;
  request_id: string;
  deployment_id: string;
  revision_id: string | null;
  model_version_id: string | null;
  api_key_id: string | null;
  batch_size: number;
  duration_ms: number;
  status: "success" | "error" | "limited";
  error_code: string | null;
  occurred_at: string;
  expires_at: string;
}

export interface InferenceRequestLogPage {
  items: InferenceRequestLog[];
  page: number;
  page_size: number;
}

export interface ModelCard {
  id: string;
  model_version_id: string;
  operational_guidance: string;
  guidance_revision: number;
  approval_status: ApprovalStatus;
  release_status: string;
  created_at: string | null;
  updated_at: string | null;
}

export interface InferenceQuery {
  since: string;
  until: string;
  page?: number;
  page_size?: number;
}

export interface ListResponse<T> {
  items: T[];
  total: number;
}

function normalizeList<T>(data: unknown): T[] {
  if (Array.isArray(data)) return data as T[];
  if (data && typeof data === "object" && Array.isArray((data as { items?: unknown }).items)) {
    return (data as { items: T[] }).items;
  }
  return [];
}

function normalizePagedList<T>(data: unknown): ListResponse<T> {
  if (Array.isArray(data)) return { items: data as T[], total: data.length };
  if (data && typeof data === "object") {
    const value = data as { items?: unknown; total?: unknown };
    if (Array.isArray(value.items) && typeof value.total === "number") {
      return { items: value.items as T[], total: value.total };
    }
  }
  return { items: [], total: 0 };
}

function commandPayload(lockVersion?: number): { expected_lock_version?: number } {
  return lockVersion === undefined ? {} : { expected_lock_version: lockVersion };
}

const METRIC_PAGE_SIZE = 200;
const LATENCY_BOUNDARIES = [5, 10, 25, 50, 100, 250, 500, 1000, 2500, 5000] as const;

function metricPercentile(histogram: Record<string, number>, percentile: number): number | null {
  const total = Object.values(histogram).reduce((sum, count) => sum + Math.max(0, count), 0);
  if (total === 0) return null;
  const threshold = Math.ceil(total * percentile);
  let seen = 0;
  for (const boundary of LATENCY_BOUNDARIES) {
    seen += Math.max(0, histogram[String(boundary)] || 0);
    if (seen >= threshold) return boundary;
  }
  return LATENCY_BOUNDARIES[LATENCY_BOUNDARIES.length - 1];
}

function summarizeMetricBuckets(items: InferenceMetricBucket[]): InferenceMetricSummary {
  const latencyBuckets: Record<string, number> = {};
  let trafficWeights: Record<string, number> = {};
  let requestCount = 0;
  let successCount = 0;
  let errorCount = 0;
  let limitedCount = 0;
  let loadFailureCount = 0;

  for (const bucket of items) {
    requestCount += bucket.request_count || 0;
    successCount += bucket.success_count || 0;
    errorCount += bucket.error_count || 0;
    limitedCount += bucket.limited_count || 0;
    loadFailureCount += bucket.load_failure_count || 0;
    for (const [boundary, count] of Object.entries(bucket.latency_buckets || {})) {
      latencyBuckets[boundary] = (latencyBuckets[boundary] || 0) + Math.max(0, count || 0);
    }
    if (Object.keys(bucket.traffic_weights || {}).length > 0) trafficWeights = bucket.traffic_weights;
  }

  return {
    request_count: requestCount,
    success_count: successCount,
    error_count: errorCount,
    limited_count: limitedCount,
    load_failure_count: loadFailureCount,
    latency_buckets: latencyBuckets,
    p50_latency_ms: metricPercentile(latencyBuckets, 0.5),
    p95_latency_ms: metricPercentile(latencyBuckets, 0.95),
    p99_latency_ms: metricPercentile(latencyBuckets, 0.99),
    traffic_weights: trafficWeights,
  };
}

export async function listRegisteredModels(projectId: string): Promise<RegisteredModel[]> {
  const response = await apiClient.get(`/projects/${projectId}/registered-models`);
  return normalizeList<RegisteredModel>(response.data);
}

export async function createRegisteredModel(
  projectId: string,
  payload: { name: string; description: string },
): Promise<RegisteredModel> {
  const response = await apiClient.post(`/projects/${projectId}/registered-models`, payload);
  return response.data;
}

export async function listModelVersions(modelId: string): Promise<ModelVersion[]> {
  const response = await apiClient.get(`/registered-models/${modelId}/versions`);
  return normalizeList<ModelVersion>(response.data);
}

export async function registerPlatformVersion(
  modelId: string,
  sourceModelLibraryId: string,
): Promise<ModelVersion> {
  const response = await apiClient.post(`/registered-models/${modelId}/versions`, {
    source_kind: "platform_joblib",
    source_model_library_id: sourceModelLibraryId,
  });
  return response.data;
}

export async function uploadOnnxArtifact(projectId: string, file: File) {
  const body = new FormData();
  body.append("file", file);
  const response = await apiClient.post(`/projects/${projectId}/model-artifacts`, body);
  return response.data as { id: string; name: string; format: string; file_size: number; sha256: string };
}

export async function registerOnnxVersion(
  modelId: string,
  payload: {
    source_artifact_id: string;
    feature_schema: FeatureField[];
    output_schema: OutputSchema;
  },
): Promise<ModelVersion> {
  const response = await apiClient.post(`/registered-models/${modelId}/versions`, {
    source_kind: "onnx_artifact",
    ...payload,
  });
  return response.data;
}

export async function approveModelVersion(versionId: string, comment = ""): Promise<ModelVersion> {
  const response = await apiClient.post(`/model-versions/${versionId}/approve`, { comment });
  return response.data;
}

export async function rejectModelVersion(versionId: string, comment: string): Promise<ModelVersion> {
  const response = await apiClient.post(`/model-versions/${versionId}/reject`, { comment });
  return response.data;
}

export async function listDeployments(projectId: string): Promise<InferenceDeployment[]> {
  const response = await apiClient.get(`/projects/${projectId}/inference-deployments`);
  return normalizeList<InferenceDeployment>(response.data);
}

export async function createDeployment(
  projectId: string,
  payload: { name: string; model_version_id: string },
): Promise<InferenceDeployment> {
  const response = await apiClient.post(`/projects/${projectId}/inference-deployments`, payload);
  return response.data;
}

export async function deleteRegisteredModel(modelId: string): Promise<{ id: string }> {
  const response = await apiClient.delete(`/registered-models/${modelId}`);
  return response.data;
}

export async function deleteDeployment(deploymentId: string): Promise<{ id: string }> {
  const response = await apiClient.delete(`/inference-deployments/${deploymentId}`);
  return response.data;
}

export async function startDeployment(deploymentId: string): Promise<InferenceDeployment> {
  const response = await apiClient.post(`/inference-deployments/${deploymentId}/start`);
  return response.data;
}

export async function stopDeployment(deploymentId: string): Promise<InferenceDeployment> {
  const response = await apiClient.post(`/inference-deployments/${deploymentId}/stop`);
  return response.data;
}

export async function predictDeployment(
  deploymentId: string,
  records: InferenceRecord[],
): Promise<PredictionResult> {
  const response = await apiClient.post(`/inference-deployments/${deploymentId}/predict`, { records });
  return response.data;
}

export async function createRollout(
  deploymentId: string,
  payload: { strategy: RolloutStrategy; targets: DeploymentTarget[]; step_schedule?: number[]; max_error_rate?: number; max_p95_ms?: number },
): Promise<DeploymentRollout> {
  const response = await apiClient.post(`/inference-deployments/${deploymentId}/rollouts`, payload);
  return response.data;
}

export async function listRollouts(deploymentId: string): Promise<ListResponse<DeploymentRollout>> {
  const response = await apiClient.get(`/inference-deployments/${deploymentId}/rollouts`);
  return normalizePagedList<DeploymentRollout>(response.data);
}

async function commandRollout(
  deploymentId: string,
  rolloutId: string,
  action: "pause" | "resume" | "rollback",
  lockVersion?: number,
): Promise<DeploymentRollout> {
  const payload = commandPayload(lockVersion);
  const response = await apiClient.post(
    `/inference-deployments/${deploymentId}/rollouts/${rolloutId}/${action}`,
    payload,
  );
  return response.data;
}

export const pauseRollout = (deploymentId: string, rolloutId: string, lockVersion?: number) =>
  commandRollout(deploymentId, rolloutId, "pause", lockVersion);

export const resumeRollout = (deploymentId: string, rolloutId: string, lockVersion?: number) =>
  commandRollout(deploymentId, rolloutId, "resume", lockVersion);

export const rollbackRollout = (deploymentId: string, rolloutId: string, lockVersion?: number) =>
  commandRollout(deploymentId, rolloutId, "rollback", lockVersion);

export async function createInferenceApiKey(
  deploymentId: string,
  payload: { scopes: ["inference.predict"]; expires_at?: string | null },
): Promise<CreatedInferenceApiKey> {
  const response = await apiClient.post(`/inference-deployments/${deploymentId}/api-keys`, payload);
  return response.data;
}

export async function listInferenceApiKeys(deploymentId: string): Promise<ListResponse<InferenceApiKey>> {
  const response = await apiClient.get(`/inference-deployments/${deploymentId}/api-keys`);
  return normalizePagedList<InferenceApiKey>(response.data);
}

export async function rotateInferenceApiKey(keyId: string): Promise<CreatedInferenceApiKey> {
  const response = await apiClient.post(`/inference-api-keys/${keyId}/rotate`);
  return response.data;
}

export async function revokeInferenceApiKey(keyId: string): Promise<InferenceApiKey> {
  const response = await apiClient.post(`/inference-api-keys/${keyId}/revoke`);
  return response.data;
}

export async function listInferenceMetrics(deploymentId: string, query: InferenceQuery): Promise<InferenceMetricPage> {
  const response = await apiClient.get(`/inference-deployments/${deploymentId}/metrics`, { params: query });
  const data = response.data as Partial<InferenceMetricPage> | null;
  return {
    items: Array.isArray(data?.items) ? data.items as InferenceMetricBucket[] : [],
    summary: (data?.summary || { request_count: 0, error_count: 0 }) as InferenceMetricSummary,
    page: typeof data?.page === "number" ? data.page : query.page || 1,
    page_size: typeof data?.page_size === "number" ? data.page_size : query.page_size || 100,
  };
}

export async function listInferenceMetricWindow(
  deploymentId: string,
  query: Pick<InferenceQuery, "since" | "until">,
): Promise<InferenceMetricPage> {
  const items: InferenceMetricBucket[] = [];
  let page = 1;

  while (true) {
    const result = await listInferenceMetrics(deploymentId, {
      ...query,
      page,
      page_size: METRIC_PAGE_SIZE,
    });
    items.push(...result.items);
    if (result.items.length < METRIC_PAGE_SIZE) break;
    page += 1;
  }

  return {
    items,
    summary: summarizeMetricBuckets(items),
    page: 1,
    page_size: METRIC_PAGE_SIZE,
  };
}

export async function listInferenceRequestLogs(deploymentId: string, query: InferenceQuery): Promise<InferenceRequestLogPage> {
  const response = await apiClient.get(`/inference-deployments/${deploymentId}/request-logs`, { params: query });
  const data = response.data as Partial<InferenceRequestLogPage> | null;
  return {
    items: Array.isArray(data?.items) ? data.items as InferenceRequestLog[] : [],
    page: typeof data?.page === "number" ? data.page : query.page || 1,
    page_size: typeof data?.page_size === "number" ? data.page_size : query.page_size || 100,
  };
}

export async function getModelCard(versionId: string): Promise<ModelCard> {
  const response = await apiClient.get(`/model-versions/${versionId}/model-card`);
  return response.data;
}

export async function updateModelCardGuidance(cardId: string, operationalGuidance: string): Promise<ModelCard> {
  const response = await apiClient.patch(`/model-cards/${cardId}/guidance`, { operational_guidance: operationalGuidance });
  return response.data;
}

export async function exportModelCard(cardId: string): Promise<Record<string, unknown>> {
  const response = await apiClient.get(`/model-cards/${cardId}/export`);
  return response.data as Record<string, unknown>;
}
